modules/nixos/glauth.nix

```diff
@@ -0,0 +1,174 @@
+            "sqlite" = "NewSQLiteHandler";
+          };
+        in
+        {
+          datastore = "plugin";
+          plugin = "${cfg.package}/bin/${dbcfg.type}.so";
+          pluginhandler = pluginHandlers.${dbcfg.type};
+          database = if (dbcfg.type != "sqlite") then (builtins.replaceStrings (singleton "\n") (singleton " ") ''
+            host=${dbcfg.host}
+            port=${toString dbcfg.port}
+            dbname=glauth
+            user=${dbcfg.username}
+            password=@db-password@
+            sslmode=${if dbcfg.ssl then "enable" else "disable"}
+          '') else "database = \"gl.db\"";
+        };
+      };
+
+
+    systemd.services.glauthPostgreSQLInit = lib.mkIf postgresCheck {
+      after = [ "postgresql.service" ];
+      before = [ "glauth.service" ];
+      bindsTo = [ "postgresql.service" ];
+      path = [ config.services.postgresql.package ];
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+        User = "postgres";
+        Group = "postgres";
+      };
+      script = ''
+        set -o errexit -o pipefail -o nounset -o errtrace
+        shopt -s inherit_errexit
+        create_role="$(mktemp)"
+        trap 'rm -f "$create_role"' ERR EXIT
+        echo "CREATE ROLE glauth WITH LOGIN PASSWORD '$(<'${dbcfg.passwordFile}')' CREATEDB" > "$create_role"
+        psql -tAc "SELECT 1 FROM pg_roles WHERE rolname='glauth'" | grep -q 1 || psql -tA --file="$create_role"
+        psql -tAc "SELECT 1 FROM pg_database WHERE datname = 'glauth'" | grep -q 1 || psql -tAc 'CREATE DATABASE "glauth" OWNER "glauth"'
+      '';
+    };
+
+    systemd.services.glauthMySQLInit = lib.mkIf mysqlCheck {
+      after = [ "mysql.service" ];
+      before = [ "glauth.service" ];
+      bindsTo = [ "mysql.service" ];
+      path = [ config.services.mysql.package ];
+      serviceConfig = {
+        Type = "oneshot";
+        RemainAfterExit = true;
+        User = config.services.mysql.user;
+        Group = config.services.mysql.group;
+      };
+      script = ''
+        set -o errexit -o pipefail -o nounset -o errtrace
+        shopt -s inherit_errexit
+        db_password="$(<'${dbcfg.passwordFile}')"
+        ( echo "CREATE USER IF NOT EXISTS 'glauth'@'localhost' IDENTIFIED BY '$db_password';"
+          echo "CREATE DATABASE glauth CHARACTER SET utf8 COLLATE utf8_unicode_ci;"
+          echo "GRANT ALL PRIVILEGES ON glauth.* TO 'glauth'@'localhost';"
+        ) | mysql -N
+      '';
+    };
+
+    users.groups.glauth = { };
+    users.users.glauth = {
+      isSystemUser = true;
+      extraGroups = singleton "glauth";
+    };
+
+    systemd.services.glauth =
+      let
+        databaseServices = attrByPath [ dbcfg.type ] [ ] {
+          "mysql" = [ "glauthMySQLInit.service" "mysql.service" ];
+          "postgres" = [ "glauthPostgreSQLInit.service" "postgresql.service" ];
+        };
+      in {
+        after = databaseServices;
+        bindsTo = databaseServices;
```

Comment:
it should probably only bind to the actual database service, the init services can just be `requires` - then I think they wouldn't need `RemainAfterExit`?